File: Data-Structure/Stack/C-Array/array_stack.c
#include "array_stack.h"

void createStack(ArrayStack **stack , int capacity) {
    (*stack) = (ArrayStack*)malloc(sizeof(ArrayStack));

    (**stack).nodes = (Node*)malloc(sizeof(Node) * capacity);

    (**stack).capacity = capacity;
    (**stack).top = -1;

}

void destroyStack(ArrayStack* stack) {
    free((*stack).nodes);
    
    free(stack);
}

void push(ArrayStack* stack, nodeDT data) {

    if ((*stack).top + 1 >= (*stack).capacity) {
        // OverFlow Error
        return;
    }

    (*stack).top++;
    (*((*stack).nodes + (*stack).top)).data = data;
}

nodeDT pop(ArrayStack* stack) {
    int top = (*stack).top;

    if (top < 0) {
        // UnderFlow
    }

    nodeDT temp;
    
    temp = (*((*stack).nodes + top)).data;

    (*stack).top--;

    return temp;
}
nodeDT top(ArrayStack* stack) {
    int top = (*stack).top;

    nodeDT temp;
    
    temp = (*((*stack).nodes + top)).data;

    return temp;
}

int getStackSize(ArrayStack* stack) {
    return (*stack).top + 1;
}

int isEmpty(ArrayStack *stack) {
    return (*stack).top == -1;
}

File: Data-Structure/Stack/Array-Based/rust-procedural/array_stack.rs

type NodeDT = i32;

pub struct Node {
    data: NodeDT,
}

pub struct ArrayStack {
    nodes: Vec<Node>,
}

pub fn create_stack(capacity: usize) -> ArrayStack {
    ArrayStack {
        nodes: Vec::with_capacity(capacity),
    }
}

pub fn push(stack: &mut ArrayStack, data: NodeDT) {
    stack.nodes.push(Node { data: data });
}

pub fn pop(stack: &mut ArrayStack) -> Option<NodeDT> {
    match stack.nodes.pop() {
        Some(node) => Some(node.data),
        None => None
    }
}

pub fn top(stack: &ArrayStack) -> Option<NodeDT> {
    match stack.nodes.last() {
        Some(node) => Some(node.data),
        None => None
    }
}

pub fn get_stack_size(stack: &ArrayStack) -> usize {
    return stack.nodes.len();
}

pub fn is_empty(stack: &ArrayStack) -> bool {
    return stack.nodes.is_empty();
}

